rasa/core/policies/unexpected_intent_policy.py

```diff
@@ -0,0 +1,914 @@
+            [
+                (intent_label, score)
+                for intent_label, score in predicted_intent_scores.items()
+            ],
+            key=lambda x: x[1],
+        )
+        query_intent_id = domain.intents.index(query_intent)
+        query_intent_similarity = similarities[0][query_intent_id]
+
+        logger.debug(
+            f"Score for intent `{query_intent}` is "
+            f"{query_intent_similarity}, while "
+            f"threshold is {self.label_thresholds[query_intent_id]}."
+        )
+        logger.debug(
+            f"Top 5 intents(in ascending order) that "
+            f"are likely here are: {sorted_intent_scores[-5:]}."
+        )
+
+        # If score for query intent is below threshold and
+        # the query intent is not the top likely intent
+        if (
+            query_intent_similarity < self.label_thresholds[query_intent_id]
+            and query_intent_id != sorted_intent_scores[-1][0]
+        ):
+            logger.debug(
+                f"Intent {query_intent}-{query_intent_id} unlikely to occur here."
+            )
+            return True
+
+        return False
+
+    @staticmethod
+    def _collect_label_id_grouped_scores(
+        output_scores: Dict[Text, Union[np.ndarray, Dict[Text, Any]]],
+        label_ids: np.ndarray,
+    ) -> Dict[int, Dict[Text, List[float]]]:
+        """Collects similarities predicted for each label id.
+
+        For each `label_id`, we collect similarity scores across
+        all trackers and categorize them into two buckets:
+            1. Similarity scores when `label_id` is the correct label.
+            2. Similarity scores when `label_id` is the wrong label.
+
+        Args:
+            output_scores: Model's predictions for each data point.
+            label_ids: Numerical IDs of labels for each data point.
+
+        Returns:
+            Both buckets of similarity scores grouped by each unique label id.
+        """
+        unique_label_ids = np.unique(label_ids).tolist()
+        if LABEL_PAD_ID in unique_label_ids:
+            unique_label_ids.remove(LABEL_PAD_ID)
+
+        label_id_scores = {
+            label_id: {POSITIVE_SCORES_KEY: [], NEGATIVE_SCORES_KEY: []}
+            for label_id in unique_label_ids
+        }
+
+        for index, all_pos_labels in enumerate(label_ids):
+
+            for candidate_label_id in unique_label_ids:
+                if candidate_label_id in all_pos_labels:
+                    label_id_scores[candidate_label_id][POSITIVE_SCORES_KEY].append(
+                        output_scores["similarities"][index, 0, candidate_label_id]
+                    )
+                else:
+                    label_id_scores[candidate_label_id][NEGATIVE_SCORES_KEY].append(
+                        output_scores["similarities"][index, 0, candidate_label_id]
+                    )
+
+        # Get only unique scores so that duplicate
+        # trackers created because of permutations are pruned out.
```

Comment:
How do we know we're not pruning away also legit duplicates? If I understand this correctly, we implicitly assume that two different trackers won't lead to the same scores. Which, I guess, is true very often but not always. Maybe worth revisiting in the future? 🤔

Reply:
Hmm, good point but I would imagine the probability of that happening really low specially with 16 point floating numbers that we use. Still I'll add a comment here to indicate the possibility and the fact that we overlook the possibility. Does that sound good to you @samsucik

Reply:
Sounds good, cheers 🙂